fix: change layout event's separator

s1syph0s opened the pull request:
#### Describe your PR, what does it fix/add?

The layout event separator is changed from "," to ";" to improve parsing ability, since some keyboard can also have "," in its name. https://github.com/Alexays/Waybar/issues/2137 https://github.com/Alexays/Waybar/pull/3224

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)

Breaks compability with waybar's hyprland/language module.

#### Is it ready for merging, or does it need work?

This is ready for merging

MahouShoujoMivutilde commented:
I think adding another event, e.g. `activelayoutv2` with `;` as a separator would better, to not break compatibility with everything else that uses it.

Original `activelayout` could be then declared deprecated in the next version, if needed, and in the meantime a note in the wiki can be added to explain why v2 is the one that should be used now.

s1syph0s commented:
Thanks for the input! I'll fix this soon

MahouShoujoMivutilde commented:
Rereading a related issue, someone here https://github.com/hyprwm/Hyprland/issues/6298#issuecomment-2212585258 said that some keyboards could contain `;` in their names. I didn't see anything like that, but sure.

Maybe let's just use `;;;`, or `|`, or something else extremely unlikely to appear in the name, but easy to split by?

Btw, see `activewindow` and `activewindowv2`, no need to EMIT_... twice, as it seems to be internal Hyprland's events, which aren't serialized in strings where splitting would matter, as opposed to the ones posted at socket2.

https://github.com/hyprwm/Hyprland/blob/4fdc0d55e4b44bb5300679025d2378fb6de0cae4/src/Compositor.cpp#L1007-L1010

or 

https://github.com/hyprwm/Hyprland/blob/4fdc0d55e4b44bb5300679025d2378fb6de0cae4/src/Compositor.cpp#L2031-L2033

vaxerski commented:
you're avoiding the problem instead of fixing it. Just replace `,`s with a space or something?

MahouShoujoMivutilde commented:
That would mean that IPC users would need to do that too, if they want to match a particular keyboard / layout, because now `.name` and `.active_keymap` won't match between `hyprctl devices` (or config) and events at socket2.

Having more unique separator character feels like a simpler change.

vaxerski commented:
`\,`?

MahouShoujoMivutilde commented:
As for escaping commas in keyboard / layouts - that would mean you still can't just split for it and get what you want (the same thing that `hyprctl devices -j` has).

```sh
❯ echo "activelayout>>some\,weird\,keyboard,Some\,Layout" | awk -F '>>' '{print $2}' | awk -F ',' '{print $2}'
weird\
```
You'll need to (pre/post)process it like e.g.

```sh
❯ echo "activelayout>>some\,weird\,keyboard,Some\,Layout" | awk -F '>>' '{print $2}' | sed 's#\\,#~#g' | awk -F ',' '{print $1}' | sed 's#~#,#g'
some,weird,keyboard

~ 
❯ echo "activelayout>>some\,weird\,keyboard,Some\,Layout" | awk -F '>>' '{print $2}' | sed 's#\\,#~#g' | awk -F ',' '{print $2}' | sed 's#~#,#g'
Some,Layout
```

As separator - it still has problem, at least for awk, 

```sh
~ 
❯ echo "activelayout>>some,weird,keyboard\,Some,Layout" | awk -F '>>' '{print $2}' | awk -F '\,' '{print $1}' 
awk: warning: escape sequence `\,' treated as plain `,'
some

~ 
❯ echo "activelayout>>some,weird,keyboard\,Some,Layout" | awk -F '>>' '{print $2}' | awk -F '\\,' '{print $1}' 
awk: warning: regexp escape sequence `\,' is not a known regexp operator
some

```

but that's fixable in awk and isn't a problem in a saner `choose`

```sh
❯ echo "activelayout>>some,weird,keyboard\,Some,Layout" | awk -F '>>' '{print $2}' | awk -F '\\\\,' '{print $1}' 
some,weird,keyboard

~ 
❯ echo "activelayout>>some,weird,keyboard\,Some,Layout" | awk -F '>>' '{print $2}' | awk -F '\\\\,' '{print $2}'
Some,Layout
```

This could work, but is worse than just unique not escaped character.

Admittedly, `||` is also not as neat as I thought it would be:

```sh
 echo "activelayout>>some,weird,keyboard||Some,Layout" | awk -F '>>' '{print $2}' | awk -F '||' '{print $1}'
some,weird,keyboard||Some,Layout

~ 
❯ echo "activelayout>>some,weird,keyboard||Some,Layout" | awk -F '>>' '{print $2}' | awk -F '||' '{print $2}'


~ 
❯ echo "activelayout>>some,weird,keyboard||Some,Layout" | awk -F '>>' '{print $2}' | awk -F '\|\|' '{print $2}' 
awk: warning: escape sequence `\|' treated as plain `|'


~
❯ echo "activelayout>>some,weird,keyboard||Some,Layout" | awk -F '>>' '{print $2}' | awk -F '\\|\\|' '{print $2}' 
Some,Layout

~ 
❯ echo "activelayout>>some,weird,keyboard||Some,Layout" | awk -F '>>' '{print $2}' | awk -F '\\|\\|' '{print $1}'
some,weird,keyboard
```

But that's, again, purely awk's problem. For waybar it wouldn't matter.

But `;;;` as a separator would work everywhere, and is more intuitive

```
❯ echo "activelayout>>some,weird,keyboard;;;Some,Layout" | awk -F '>>' '{print $2}' | awk -F ';;;' '{print $1}' 
some,weird,keyboard

~ 
❯ echo "activelayout>>some,weird,keyboard;;;Some,Layout" | awk -F '>>' '{print $2}' | awk -F ';;;' '{print $2}'
Some,Layout
```

You get what you split for, no need to learn any escaping.

s1syph0s commented:
Should I just replace the separator from `||` to `;;`?

cnt0 commented:
I'd propose something like
```
activalayoutv2>>KEYBOARDNAME,LAYOUTNAME,LAYOUTNAMEPOS
```
where integer `LAYOUTNAMEPOS` points to the beginning of `LAYOUTNAME` in the event string.

This allows us to 1) correctly parse the event regardless of layout and keyboard name contents 2) maintains the backwards compatibility with `activelayout`, so we actually don't need to introduce the new event at all.

s1syph0s commented:
I think that we still need to introduce a new event `activelayoutv2`, since adding `LAYOUTNAMEPOS` at the end will break waybar's current parser..

vaxerski commented:
> This could work, but is worse than just unique not escaped character.

dumb statement, what is there to make you so sure there is any character that will ever be unique?

MahouShoujoMivutilde commented:
Practical experience of how humans make up names for things meant to be read by other humans? 

`;` is rather common part of written language, so to see it in some `Gaymer Keyboard (uwu edition; international)` is more likely than `;;;` or `||` or a `\t` (tab). 

Maybe someone who has a keyboard with a fully programmable controller in it can mess this up.

But if you expect a completely arbitrary text in the name and layout, sure, escaping the separator character is the way. 

If we're being this pedantic, escaping the separator during the serialization and unescaping on user side should **always** be done, for all events with arbitrary output, and maybe all for the sake of consistency / simpler maintainability should the expected text change later.

![12-08-2024-00_45_08](https://github.com/user-attachments/assets/ac937069-caab-4d25-b5f7-c974fc84cd02)

I just think that the less unescaping the user needs to do, the better.

vaxerski commented:
escaping is a common practice and I still believe `\,` is the way

s1syph0s commented:
![screenshot_12-08-2024-194517](https://github.com/user-attachments/assets/f82ab5bf-7de9-4e17-9204-27e0fa7846d5)

I tried to use `\,` as the separator, but it is an unknown escape sequence. If I build this and gets the event output, I will receive this, which has no difference:

```
activelayout>>wl_keyboard,English (US)
activelayoutv2>>wl_keyboard,English (US)
```

If I remove the `v2` for the second output and put both in a diffcheck tool, the tool won't find any difference.

![screenshot_12-08-2024-200146](https://github.com/user-attachments/assets/1b43297a-fdd1-40f1-9256-56ea753a424b)

Using rust to parse the event output, trying to read `\,` will result in an error, since the escape sequence is not known. Do you mean to use a literal `\,`, which leads to `activelayoutv2>>wl_keyboard\,English (US)`, or escape the comma?

vaxerski commented:
you need to put `\\,` in the source

s1syph0s commented:
Hi, is there any update to this issue? I'll be happy to adjust the PR

vaxerski commented:
well I am still convinced escaping the commas with `\` is the way to go.

s1syph0s commented:
I created a [commit](https://github.com/hyprwm/Hyprland/pull/7275/commits/3b55c64c2978cafcb20e26b73f9795c3c8b68c4d) that escapes the comma. Could you review the code?

vaxerski commented:
it's the opposite of what we want. We want to escape the commas in the layout name

s1syph0s commented:
Hi,

sorry for the late response.. I was taking a break from software dev in my free time..

I tried escaping the comma in both the layout and the language, but the event still can't be parsed properly..

```
raw: activelayoutv2>>hp\,-inc-hyperx-alloy-origins,German (no dead keys)
split: hp\ || -inc-hyperx-alloy-origins,German (no dead keys)

raw: activelayoutv2>>hp\,-inc-hyperx-alloy-origins,English (US)
split: hp\ || -inc-hyperx-alloy-origins,English (US)
```

Another idea that I would propose is to remove the comma completely from the keyboard name and the layout, but this is also not backwards compatible. To fix this in waybar for example, we also need to remove the commas from the input keyboard name. What do you think about this?

vaxerski commented:
closing due to inactivity
